File: core/cyp/backtest/optimize.py
"""扫参择优：对一组 StrategyConfig 跑回测，按目标函数排序，选出最优策略配置。

三档统一下，回测用的就是实盘管线——扫出的最优配置可直接注入 Orchestrator(strategy=...)。
"""

from __future__ import annotations

import itertools
from typing import Callable

from pydantic import BaseModel

from cyp.agents import StrategyConfig
from cyp.backtest.engine import Backtester
from cyp.backtest.pbo import pbo
from cyp.backtest.stats import deflated_sharpe, sharpe
from cyp.config import Settings
from cyp.contracts import Candle


class SweepResult(BaseModel):
    config: StrategyConfig
    metrics: dict
    score: float


def default_objective(m: dict) -> float:
    """奖励收益、惩罚回撤（简单可解释）。"""
    return round(m["total_return"] - m["max_drawdown"], 4)


def grid(**options) -> list[StrategyConfig]:
    """笛卡尔积生成候选配置：grid(enter_threshold=[...], k_stop=[...], ...)。"""
    keys = list(options)
    return [StrategyConfig(**dict(zip(keys, combo)))
            for combo in itertools.product(*(options[k] for k in keys))]


async def sweep(settings: Settings, symbol: str, candles: list[Candle],
                configs: list[StrategyConfig], window: int = 60,
                objective: Callable[[dict], float] = default_objective) -> list[SweepResult]:
    results: list[SweepResult] = []
    for cfg in configs:
        report = await Backtester(settings, symbol, candles, window, strategy=cfg).run()
        results.append(SweepResult(config=cfg, metrics=report.metrics, score=objective(report.metrics)))
    results.sort(key=lambda r: r.score, reverse=True)
    return results


def _bar_returns(equity_curve: list[float]) -> list[float]:
    return [equity_curve[i] / equity_curve[i - 1] - 1
            for i in range(1, len(equity_curve)) if equity_curve[i - 1] > 0]


class RobustResult(BaseModel):
    best_config: StrategyConfig
    is_metrics: dict            # 样本内（仅参考）
    oos_metrics: dict           # 样本外（真相）
    pbo: float                  # 过拟合概率（越低越好）
    deflated_sharpe: float      # 扣除多试验运气后的夏普显著性
    verdict: str                # PASS / REJECT(疑似过拟合)


async def robust_sweep(settings: Settings, symbol: str, candles: list[Candle],
                       configs: list[StrategyConfig], window: int = 60, oos_frac: float = 0.3,
                       objective: Callable[[dict], float] = default_objective,
                       pbo_max: float = 0.5, dsr_min: float = 0.5) -> RobustResult:
    """诚实择优：样本内(IS)挑最优 → 样本外(OOS)验证 + PBO + Deflated Sharpe 出裁决。

    不再"样本内挑最高分"——那是过拟合。IS 只用来选，OOS 才算数。
    """
    n = len(candles)
    split = max(window + 2, int(n * (1 - oos_frac)))
    is_candles = candles[:split]
    oos_candles = candles[split - window:]      # 保留 window 预热

    is_reports = [await Backtester(settings, symbol, is_candles, window, strategy=c).run() for c in configs]
    is_scores = [objective(r.metrics) for r in is_reports]
    is_curves = [_bar_returns(r.equity_curve) for r in is_reports]
    best_i = max(range(len(configs)), key=lambda i: is_scores[i])

    usable = min((len(c) for c in is_curves if len(c) > 1), default=0)
    matrix = [c[:usable] for c in is_curves if len(c) >= usable > 1]
    pbo_val = pbo(matrix, s=6) if usable >= 12 and len(matrix) >= 2 else 0.0
    trial_sharpes = [sharpe(c) for c in matrix] or [0.0]

    oos_rep = await Backtester(settings, symbol, oos_candles, window, strategy=configs[best_i]).run()
    oos_ret = _bar_returns(oos_rep.equity_curve)
    dsr = deflated_sharpe(oos_ret, trial_sharpes) if len(oos_ret) > 1 else 0.0

    passed = pbo_val <= pbo_max and oos_rep.metrics["total_return"] > 0 and dsr >= dsr_min
    return RobustResult(
        best_config=configs[best_i], is_metrics=is_reports[best_i].metrics,
        oos_metrics=oos_rep.metrics, pbo=round(pbo_val, 4), deflated_sharpe=round(dsr, 4),
        verdict="PASS" if passed else "REJECT(疑似过拟合)",
    )

File: internal/backtest/optimize.go
package backtest

import (
	"errors"
	"math"
	"sort"
)

type StrategyConfig struct {
	EnterThreshold        float64 `json:"enter_threshold"`
	StopVolMultiple       float64 `json:"k_stop"`
	TakeProfitVolMultiple float64 `json:"k_tp"`
}

func DefaultStrategyConfig() StrategyConfig { return StrategyConfig{} }

// Grid produces a Cartesian product. Empty dimensions use the default value.
func Grid(enterThresholds, stopMultiples, takeProfitMultiples []float64) []StrategyConfig {
	if len(enterThresholds) == 0 {
		enterThresholds = []float64{0}
	}
	if len(stopMultiples) == 0 {
		stopMultiples = []float64{0}
	}
	if len(takeProfitMultiples) == 0 {
		takeProfitMultiples = []float64{0}
	}
	result := make([]StrategyConfig, 0, len(enterThresholds)*len(stopMultiples)*len(takeProfitMultiples))
	for _, threshold := range enterThresholds {
		for _, stop := range stopMultiples {
			for _, takeProfit := range takeProfitMultiples {
				result = append(result, StrategyConfig{
					EnterThreshold: threshold, StopVolMultiple: stop,
					TakeProfitVolMultiple: takeProfit,
				})
			}
		}
	}
	return result
}

type Objective func(Metrics) float64

func DefaultObjective(metrics Metrics) float64 {
	return roundTo(metrics.TotalReturn-metrics.MaxDrawdown, 4)
}

type SweepResult struct {
	Config  StrategyConfig `json:"config"`
	Metrics Metrics        `json:"metrics"`
	Score   float64        `json:"score"`
	Lessons []string       `json:"lessons"`
}

func Sweep(params Params, configs []StrategyConfig, objective Objective) ([]SweepResult, error) {
	if len(configs) == 0 {
		return nil, errors.New("at least one strategy config is required")
	}
	if objective == nil {
		objective = DefaultObjective
	}
	results := make([]SweepResult, 0, len(configs))
	for _, strategy := range configs {
		report, err := RunWithStrategy(params, strategy)
		if err != nil {
			return nil, err
		}
		lessons := append([]string{}, report.Lessons...)
		if len(lessons) > 5 {
			lessons = lessons[len(lessons)-5:]
		}
		results = append(results, SweepResult{
			Config: strategy, Metrics: report.Metrics,
			Score: objective(report.Metrics), Lessons: lessons,
		})
	}
	sort.SliceStable(results, func(i, j int) bool { return results[i].Score > results[j].Score })
	return results, nil
}

type RobustResult struct {
	BestConfig     StrategyConfig `json:"best_config"`
	InSample       Metrics        `json:"is_metrics"`
	OutOfSample    Metrics        `json:"oos_metrics"`
	PBO            float64        `json:"pbo"`
	DeflatedSharpe float64        `json:"deflated_sharpe"`
	Verdict        string         `json:"verdict"`
}

func RobustSweep(
	params Params,
	configs []StrategyConfig,
	outOfSampleFraction float64,
	pboMaximum, dsrMinimum float64,
) (RobustResult, error) {
	if len(configs) == 0 {
		return RobustResult{}, errors.New("at least one strategy config is required")
	}
	if outOfSampleFraction <= 0 || outOfSampleFraction >= 1 {
		outOfSampleFraction = 0.3
	}
	split := int(float64(params.Bars) * (1 - outOfSampleFraction))
	if split < params.Window+2 {
		split = params.Window + 2
	}
	if split >= params.Bars {
		return RobustResult{}, errors.New("not enough bars for out-of-sample validation")
	}
	inParams := params
	inParams.Bars = split
	inReports := make([]Report, len(configs))
	scores := make([]float64, len(configs))
	returnMatrix := make([][]float64, len(configs))
	best := 0
	for index, strategy := range configs {
		report, err := RunWithStrategy(inParams, strategy)
		if err != nil {
			return RobustResult{}, err
		}
		inReports[index] = report
		scores[index] = DefaultObjective(report.Metrics)
		returnMatrix[index] = BarReturns(report.EquityCurve)
		if scores[index] > scores[best] {
			best = index
		}
	}
	usable := len(returnMatrix[0])
	for _, returns := range returnMatrix {
		if len(returns) < usable {
			usable = len(returns)
		}
	}
	matrix := make([][]float64, 0, len(returnMatrix))
	trialSharpes := make([]float64, 0, len(returnMatrix))
	if usable > 1 {
		for _, returns := range returnMatrix {
			trimmed := append([]float64{}, returns[:usable]...)
			matrix = append(matrix, trimmed)
			trialSharpes = append(trialSharpes, Sharpe(trimmed))
		}
	}
	pboValue := 0.0
	if usable >= 12 && len(matrix) >= 2 {
		pboValue = ProbabilityBacktestOverfit(matrix, 6, nil)
	}
	if len(trialSharpes) == 0 {
		trialSharpes = []float64{0}
	}

	outParams := params
	outParams.Bars = params.Bars - split + params.Window
	outParams.Seed = params.Seed + int64(split) + 1
	outReport, err := RunWithStrategy(outParams, configs[best])
	if err != nil {
		return RobustResult{}, err
	}
	outReturns := BarReturns(outReport.EquityCurve)
	dsr := 0.0
	if len(outReturns) > 1 {
		dsr = DeflatedSharpe(outReturns, trialSharpes)
	}
	verdict := "REJECT(疑似过拟合)"
	if pboValue <= pboMaximum && outReport.Metrics.TotalReturn > 0 && dsr >= dsrMinimum {
		verdict = "PASS"
	}
	return RobustResult{
		BestConfig: configs[best], InSample: inReports[best].Metrics,
		OutOfSample: outReport.Metrics, PBO: roundTo(pboValue, 4),
		DeflatedSharpe: roundTo(dsr, 4), Verdict: verdict,
	}, nil
}

func roundTo(value float64, precision int) float64 {
	power := math.Pow10(precision)
	return math.Round(value*power) / power
}

func BarReturns(equityCurve []float64) []float64 {
	result := make([]float64, 0, len(equityCurve)-1)
	for index := 1; index < len(equityCurve); index++ {
		if equityCurve[index-1] > 0 {
			result = append(result, equityCurve[index]/equityCurve[index-1]-1)
		}
	}
	return result
}
